src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,209 @@
+    // solves?
+    VarsType operand_{};
+    while (not convergence_reason_) {
+      operand_ = source - linear_operator(result);
+      const double initial_residual_magnitude =
+          sqrt(inner_product(operand_, operand_));
+      convergence_reason_ = Convergence::criteria_match(
+          convergence_criteria_, iteration_, initial_residual_magnitude,
+          initial_residual_magnitude);
+      if (convergence_reason_) {
+        break;
+      }
+      operand_ /= initial_residual_magnitude;
+      basis_history_[0] = operand_;
+
+      std::pair<DenseVector<double>, double> minres_and_magnitude{};
+      for (size_t k = 0; k < restart_; k++) {
+        if (use_preconditioner) {
+          preconditioned_basis_history_[k] = preconditioner(operand_);
+        }
+        operand_ = linear_operator(
+            use_preconditioner ? preconditioned_basis_history_[k] : operand_);
+        // Resize matrix
+        orthogonalization_history_.resize(k + 2, k + 1);
+        // Make sure the new entries are zero
+        for (size_t i = 0; i < orthogonalization_history_.rows(); i++) {
+          orthogonalization_history_(
+              i, orthogonalization_history_.columns() - 1) = 0.;
+        }
+        for (size_t j = 0; j < orthogonalization_history_.columns(); j++) {
+          orthogonalization_history_(orthogonalization_history_.rows() - 1, j) =
+              0.;
+        }
+        // Arnoldi orthogonalization
+        for (size_t j = 0; j < k + 1; j++) {
+          const double orthogonalization =
+              inner_product(basis_history_[j], operand_);
+          orthogonalization_history_(j, k) = orthogonalization;
+          operand_ -= orthogonalization * basis_history_[j];
+        }
+        orthogonalization_history_(k + 1, k) =
+            sqrt(inner_product(operand_, operand_));
+        // Least-squares solve for the minimal residual
+        minres_and_magnitude = gmres_detail::minimal_residual(
+            orthogonalization_history_, initial_residual_magnitude);
+        iteration_++;
+        convergence_reason_ = Convergence::criteria_match(
+            convergence_criteria_, iteration_, minres_and_magnitude.second,
+            initial_residual_magnitude);
+        if (UNLIKELY(convergence_reason_)) {
+          break;
+        } else if (k + 1 < restart_) {
+          operand_ /= orthogonalization_history_(k + 1, k);
+          basis_history_[k + 1] = operand_;
+        }
+      }
+      for (size_t i = 0; i < minres_and_magnitude.first.size(); i++) {
+        result += minres_and_magnitude.first[i] *
+                  gsl::at(use_preconditioner ? preconditioned_basis_history_
+                                             : basis_history_,
+                          i);
+      }
+    }
+    return result;
+  }
+
+ private:
+  Convergence::Criteria convergence_criteria_{};
+  ::Verbosity verbosity_{::Verbosity::Verbose};
+  size_t restart_{};
+
+  mutable DenseMatrix<double> orthogonalization_history_{};
```

Comment:
`orthogonalization_history` and `basis_history` are challenging names for these objects, I feel. Could you give them names that more closely reflect their mathematical foundations?

Reply:
hm do you have ideas? I wanted to give them names that reflect their role in the algorithm. I use the same names in the parallel version in `ParallelAlgorithms/LinearSolver/`

Reply:
I found `history` to be confusing in absence of hyperbolic evolution. For the basis vectors, perhaps `orthonormal_basis` or `krylov_basis`? And for the `orthogonalization_history` perhaps something hinting at the fact it is a matrix of inner products of the basis vectors evaluated according to the given operator? My main point is that `history` sounds like only the final entry is representative of the current iteration, where these objects iteratively build a basis that spans a larger and larger subspace.

Reply:
I see how `history` doesn't work well for those variables that update during the iterations, i.e. that are not merely appended-to. I can think of names for all but `orthogonalization_history` and `residual_history`. The former is built iteratively from inner products between existing and potential basis vectors and then Givens-rotated to become upper-triangular. The latter holds the remaining residual in its last entry, but and the other entries "source" the minimum residual vector as `g` in `R * y = g` where `R` is the upper-triangular matrix. If we come up with good names I'm happy to change, else propose to just go ahead unless you have strong objections to a particular name.

Reply:
Alright, fair enough. I am at a loss for better variable names, perhaps just add a couple of brief code comments effectively with your above explanation about what each holds nearby the member variable declarations? You can also squash that directly if you like.

Reply:
Added a bunch of comments to the variables